### ISSUE TITLE
Cannot parse object as JSON

### ISSUE THREAD

**Issue by AlbertDeFusco** [jupyter-incubator/sparkmagic#558]
I saw that this was fixed in version 0.12.9 but I'm still having trouble.

Local package versions:
* Python 3.7.3
* Sparkmagic: 0.12.9
* Pandas: 0.24.2

Remote package versions:
* Python: 3.7.0
* PySpark: 2.4.0

Here's my notebook

```python
# Cell 1
age = spark.createDataFrame(['10','11','13'], 'string').toDF('age')

# Cell 2
%%spark -o a
```

I get the following error. If I execute Cell 2 **twice** then the `age` object is returned to `%%local` correctly as a Pandas DataFrame.

> An internal error was encountered.
Please file an issue at https://github.com/jupyter-incubator/sparkmagic
Error:
Cannot parse object as JSON: '['{"age":"10"}', '{"age":"11"}', '{"age":"13"}', '----------------------------------------', "Exception happened during processing of request from ('127.0.0.1', 59772)", 'Traceback (most recent call last):', '  File "/opt/anaconda3/lib/python3.7/socketserver.py", line 313, in _handle_request_noblock', '    self.process_request(request, client_address)', '  File "/opt/anaconda3/lib/python3.7/socketserver.py", line 344, in process_request', '    self.finish_request(request, client_address)', '  File "/opt/anaconda3/lib/python3.7/socketserver.py", line 357, in finish_request', '    self.RequestHandlerClass(request, client_address, self)', '  File "/opt/anaconda3/lib/python3.7/socketserver.py", line 717, in __init__', '    self.handle()', '  File "/mnt/yarn/usercache/instructor/appcache/application_1559921466801_0213/container_1559921466801_0213_01_000001/pyspark.zip/pyspark/accumulators.py", line 266, in handle', '    poll(authenticate_and_accum_updates)', '  File "/mnt/yarn/usercache/instructor/appcache/application_1559921466801_0213/container_1559921466801_0213_01_000001/pyspark.zip/pyspark/accumulators.py", line 241, in poll', '    if func():', '  File "/mnt/yarn/usercache/instructor/appcache/application_1559921466801_0213/container_1559921466801_0213_01_000001/pyspark.zip/pyspark/accumulators.py", line 254, in authenticate_and_accum_updates', '    received_token = self.rfile.read(len(auth_token))', "TypeError: object of type 'NoneType' has no len()", '----------------------------------------']'

**Comment by itamarst** [comment]
Thanks for the bug report! I will try to take a look at this when I'm back from vacation, in September, if someone else doesn't get to it first.

**Comment by itamarst** [comment]
I can't reproduce with the current Docker image (Spark 2.3.3, Python 3.5). Will try to update to newer versions I guess.

**Comment by AlbertDeFusco** [comment]
Thanks. Further, my cluster is an AWS EMR

**Release label** :emr-5.23.0
**Hadoop distribution** :Amazon 2.8.5
**Applications** :Spark 2.4.0, Hive 2.3.4, Livy 0.5.0, Hue 4.3.0

**Comment by devender-yadav** [comment]
Working fine on EMR 5.26

**Comment by itamarst** [comment]
@AlbertDeFusco can you retry with newer EMR?

**Comment by nileshlange** [comment]
Hi,

I am also getting same issue in our environment. We are not using EMR we have livy and spark magic with Cloudera cluster.

An internal error was encountered.
Please file an issue at https://github.com/jupyter-incubator/sparkmagic
Error:
Cannot parse object as JSON: '['b\'{"database":"default","tableName":"*************","isTemporary":false}\'', 'b\'{"database":"default","tableName":"******************","isTemporary":false}\'', 'b\'{"database":"default","tableName":"*******************","isTemporary":false}\'', '----------------------------------------', "Exception happened during processing of request from ('127.0.0.1', 11642)", 'Traceback (most recent call last):', '  File "/opt/cloudera/parcels/Anaconda-5.1.0.1/lib/python3.6/socketserver.py", line 317, in _handle_request_noblock', '    self.process_request(request, client_address)', '  File "/opt/cloudera/parcels/Anaconda-5.1.0.1/lib/python3.6/socketserver.py", line 348, in process_request', '    self.finish_request(request, client_address)', '  File "/opt/cloudera/parcels/Anaconda-5.1.0.1/lib/python3.6/socketserver.py", line 361, in finish_request', '    self.RequestHandlerClass(request, client_address, self)', '  File "/opt/cloudera/parcels/Anaconda-5.1.0.1/lib/python3.6/socketserver.py", line 696, in __init__', '    self.handle()', '  File "/opt/cloudera/parcels/SPARK2-2.3.0.cloudera4-1.cdh5.13.3.p0.611179/lib/spark2/python/lib/pyspark.zip/pyspark/accumulators.py", line 263, in handle', '    poll(authenticate_and_accum_updates)', '  File "/opt/cloudera/parcels/SPARK2-2.3.0.cloudera4-1.cdh5.13.3.p0.611179/lib/spark2/python/lib/pyspark.zip/pyspark/accumulators.py", line 238, in poll', '    if func():', '  File "/opt/cloudera/parcels/SPARK2-2.3.0.cloudera4-1.cdh5.13.3.p0.611179/lib/spark2/python/lib/pyspark.zip/pyspark/accumulators.py", line 251, in authenticate_and_accum_updates', '    received_token = self.rfile.read(len(auth_token))', "TypeError: object of type 'NoneType' has no len()", '----------------------------------------']'

**Comment by itamarst** [comment]
@nileshlange what version sparkmagic?

**Comment by nileshlange** [comment]
Hi itamarst, thank you for the quick response. do you know how to check the version of sparkmagic from jupyter notebook / terminal ?

**Comment by itamarst** [comment]
```python
import sparkmagic
print(sparkmagic.__version__)
```

In a Python prompt (or Jupyter).

**Comment by itamarst** [comment]
Ah. Can you try upgrading to 0.12.9?

**Comment by nileshlange** [comment]
Sure i will try and let you know the result. Thank you :)

**Comment by AlbertDeFusco** [comment]
I'll spin up a new EMR version and report back soon.

**Comment by xinofekuator** [comment]
I was having this issue with JupyterLab but it is not happening with Jupyter Notebook (so I changed Jupyter Lab for Jupyter notebook and now it is not happening).

Version used:

```
jupyter==1.0.0
jupyter-client==5.3.1
jupyter-console==6.0.0
jupyter-core==4.5.0
jupyterlab==1.0.9
jupyterlab-server==1.0.6

```

**Comment by nileshlange** [comment]
@itamarst we have updated sparkmagic version still we are getting same issue.

**Comment by itamarst** [comment]
OK, thanks for checking. Looking deeper into the traceback, this seems like a PySpark issue. 

And indeed I found https://issues.apache.org/jira/browse/SPARK-26019

Sounds like this is fixed in Spark 2.3.3 and 2.4.1 (or later), so try upgrading to one of those.
